cmd/snap/cmd_run.go

```diff
@@ -1013,3 +1062,256 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
 		return syscallExec(cmd[0], cmd, envForExec(nil))
 	}
 }
+
+var errDBusUnknownMethod = errors.New("org.freedesktop.DBus.Error.UnknownMethod")
+var errDBusNameHasNoOwner = errors.New("org.freedesktop.DBus.Error.NameHasNoOwner")
+var errDBusSpawnChildExited = errors.New("org.freedesktop.DBus.Error.Spawn.ChildExited")
+
+// doCreateTransientScope creates a systemd transient scope with specified properties.
+//
+// The scope is created by asking systemd via the specified DBus connection.
+// The unit name and the PID to attach are provided as well. The DBus method
+// call is performed outside confinement established by snap-confine.
+func doCreateTransientScope(conn *dbus.Conn, unitName string, pid int) error {
+	// The property and auxUnit types are not well documented but can be traced
+	// from systemd source code. Systemd defines the signature of
+	// StartTransientUnit as "ssa(sv)a(sa(sv))". The signature can be
+	// decomposed as follows:
+	//
+	// Partial documentation, at the time of this writing, is available at
+	// https://www.freedesktop.org/wiki/Software/systemd/dbus/
+	//
+	// unitName string // name of the unit to start
+	// jobMode string  // corresponds to --job-mode= (see systemctl(1) manual page)
+	// properties []struct{
+	//   Name string
+	//   Value interface{}
+	// } // properties describe properties of the started unit
+	// auxUnits []struct {
+	//   Name string
+	//   Properties []struct{
+	//   	Name string
+	//   	Value interface{}
+	//	 }
+	// } // auxUnits describe any additional units to define.
+	type property struct {
+		Name string
+		// XXX: This is getting marshaled as an invalid variant.
+		Value interface{}
+	}
+	type auxUnit struct {
+		Name  string
+		Props []property
+	}
+
+	mode := "fail"
+	properties := []property{{"PIDs", []uint{uint(pid)}}}
+	aux := []auxUnit(nil)
+	systemd := conn.Object("org.freedesktop.systemd1", "/org/freedesktop/systemd1")
+	call := systemd.Call(
+		"org.freedesktop.systemd1.Manager.StartTransientUnit",
+		0,
+		unitName,
+		mode,
+		properties,
+		aux,
+	)
+	var job dbus.ObjectPath
+	if err := call.Store(&job); err != nil {
+		if dbusErr, ok := err.(dbus.Error); ok {
+			logger.Debugf("StartTransientUnit failed with %q: %v", dbusErr.Name, dbusErr.Body)
```

Comment:
all this complex error handling doesn't seem to be unit tested

Reply:
That's correct, only the happy path is tested. I this branch developed the infrastructure to write such tests but as it was huge I had to stop somewhere. I do plan to get the test coverage up significantly, just didn't want to do it here.

Reply:
This is now all extensively tested.